exposed/src/main/kotlin/org/jetbrains/exposed/sql/transactions/ThreadLocalTransactionManager.kt

```diff
@@ -7,138 +7,222 @@ import org.jetbrains.exposed.sql.Transaction
+    override fun newTransaction(isolation: Int, outerTransaction: Transaction?): Transaction = Transaction(
+        ThreadLocalTransaction(
+            db = db,
+            transactionIsolation = if (outerTransaction != null) outerTransaction.transactionIsolation else isolation,
+            threadLocal = threadLocal,
+            outerTransaction = outerTransaction,
+            connectionProvider = {
+                if (outerTransaction == null) {
+                    db.connector().apply {
+                        autoCommit = false
+                        this.transactionIsolation = isolation
+                    }
+                } else {
+                    outerTransaction.connection
+                }
+            }
+        )
+    ).apply {
         threadLocal.set(this)
     }
 
     override fun currentOrNull(): Transaction? = threadLocal.get()
 
-    private class ThreadLocalTransaction(override val db: Database, isolation: Int, val threadLocal: ThreadLocal<Transaction>) : TransactionInterface {
+    private class ThreadLocalTransaction(
+        override val db: Database,
+        override val transactionIsolation: Int,
+        val threadLocal: ThreadLocal<Transaction>,
+        override val outerTransaction: Transaction?,
+        private val connectionProvider: () -> Connection = { db.connector() }
+    ) : TransactionInterface {
 
-        private val connectionLazy = lazy(LazyThreadSafetyMode.NONE) {
-            db.connector().apply {
-                autoCommit = false
-                transactionIsolation = isolation
-            }
-        }
+        private val connectionLazy = lazy(LazyThreadSafetyMode.NONE) { connectionProvider() }
         override val connection: Connection
             get() = connectionLazy.value
 
-        override val outerTransaction: Transaction? = threadLocal.get()
+        private val topLevelTransaction = outerTransaction == null
+        private var savepoint: Savepoint? = if (!topLevelTransaction) {
+            connection.setSavepoint("POINT_$name")
+        } else null
+
 
         override fun commit() {
-            if (connectionLazy.isInitialized())
-                connection.commit()
+            if (connectionLazy.isInitialized()) {
+                if (topLevelTransaction) {
+                    connection.commit()
+                } else {
+                    // do nothing in nested. close() will commit everything and release savepoint.
+                }
+            }
         }
 
         override fun rollback() {
             if (connectionLazy.isInitialized() && !connection.isClosed) {
-                connection.rollback()
+                if (topLevelTransaction) {
+                    connection.rollback()
+                } else {
+                    connection.rollback(savepoint)
+                    savepoint = connection.setSavepoint("POINT_$name")
```

Comment:
Looks like it's unnecessary to reset savepoint because rollback doesn't release one. Anyway `NestedTransactionTest` was passed.